Add dataset deserializer for 'datatype'

simonbray opened the pull request:
The purpose of this change (suggested by @nsoranzo) is to allow datatypes to be changed via the API/bioblend.

I see that being able to update a datatype for a running job might be an issue, so please let me know if another approach would be better.

Fixes #4363.

simonbray committed:
add extension deserializer

nsoranzo commented:
It would be good to have an API test for that, I think.

simonbray committed:
add test

simonbray committed:
fix test

simonbray commented:
Added a test.

mvdbeek commented on lib/galaxy_test/api/test_datasets.py:
```diff
@@ -136,3 +136,16 @@ def test_tag_change(self):
+        updated_hda = self._put(
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id),
+            update_payload).json()
+        assert updated_hda['extension'] == 'dat'
```
I think this shows it does not work, it doesn't update the metadata (probably just sets the extension in the database), and dat is not a valid extension, so this must fail.
For an implementation you might want to look at https://github.com/mvdbeek/galaxy/blob/cd7a2c3418d6540ab3bb3177b0b0216c54ebff56/lib/galaxy/webapps/galaxy/controllers/dataset.py#L365.

A test should then also demonstrate that datatype specific metadata has been set.

nsoranzo replied:
Ops, that looked too easy! @simonbray You can probably move the implementation that Marius linked to a new `DatasetAssociationDeserializer.deserialize_extension()` .

simonbray committed:
Merge remote-tracking branch 'upstream/dev' into HEAD

simonbray committed:
redo datatype deserializer

simonbray committed:
add error when datatype cannot be updated

simonbray committed:
lint...

nsoranzo commented on lib/galaxy/managers/datasets.py:
```diff
@@ -684,6 +686,17 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
 
+    def deserialize_datatype(self, item, key, val, **context):
+        if item.datatype.allow_datatype_change and context['trans'].app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            if managers.hdas.HDAManager(self.app).ok_to_edit_metadata(item.dataset_id):
```
```suggestion
            if DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
```

Or `ok_to_edit_metadata()` should actually be a method of the `DatasetInstance` class in `lib/galaxy/model/__init__.py`, I think.

nsoranzo commented on lib/galaxy/managers/datasets.py:
```diff
@@ -684,6 +686,17 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
         return unwrapped_val
 
+    def deserialize_datatype(self, item, key, val, **context):
+        if item.datatype.allow_datatype_change and context['trans'].app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
```
```suggestion
        if item.datatype.allow_datatype_change and self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
```

nsoranzo commented on lib/galaxy/managers/datasets.py:
```diff
@@ -684,6 +686,17 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
+    def deserialize_datatype(self, item, key, val, **context):
+        if item.datatype.allow_datatype_change and context['trans'].app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            if managers.hdas.HDAManager(self.app).ok_to_edit_metadata(item.dataset_id):
+                context['trans'].app.datatypes_registry.change_datatype(item, val)
```
```suggestion
                item.change_datatype(val)
```

nsoranzo commented on lib/galaxy/managers/datasets.py:
```diff
@@ -684,6 +686,17 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
+                return item.datatype
+        context['trans'].response.status = 500
+        context['trans'].error_message = 'Datatype could not be updated'
+        return False
```
Should raise an exception instead.

simonbray committed:
fix some suggestions from @nsoranzo

nsoranzo commented on lib/galaxy/managers/datasets.py:
```diff
@@ -684,6 +685,17 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
+                self.app.datatypes_registry.set_external_metadata_tool.tool_action.execute(self.app.datatypes_registry.set_external_metadata_tool, context['trans'], incoming={'input1': item}, overwrite=False)  # overwrite is False as per existing behavior
+                return item.datatype
+        context['trans'].response.status = 500
+        context['trans'].error_message = 'Datatype could not be updated.'
```
```suggestion
```

nsoranzo commented on lib/galaxy/managers/datasets.py:
```diff
@@ -684,6 +685,17 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
+    def deserialize_datatype(self, item, key, val, **context):
+        if item.datatype.allow_datatype_change and self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            if DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
+                item.change_datatype(val)
```
To return more specific error messages, I would restructure this as:
```python
        if not item.datatype.allow_datatype_change:
        	raise SomeException("message")
        if not self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
        	raise SomeException("message")
        if not DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
        	raise AnotherException("message")
        item.change_datatype(val)
        ...
```

simonbray committed:
add multiple error messages

nsoranzo commented on lib/galaxy/managers/datasets.py:
```diff
@@ -686,15 +686,16 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
+        if not item.datatype.allow_datatype_change:
+            raise Exception("The current datatype does not allow datatype changes.")
+        if not self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            raise Exception("Either the target datatype does not exist, or it does not allow datatype changes.")
```
```suggestion
		if not self.app.datatypes_registry.get_datatype_by_extension(val):
			raise RequestParameterInvalidException("The target datatype does not exist.")
        if not self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
            raise RequestParameterInvalidException("The target datatype does not allow datatype changes.")
```

nsoranzo commented on lib/galaxy/managers/datasets.py:
```diff
@@ -686,15 +686,16 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
-        context['trans'].error_message = 'Datatype could not be updated.'
-        raise Exception('Datatype could not be updated.')
+        if not item.datatype.allow_datatype_change:
+            raise Exception("The current datatype does not allow datatype changes.")
```
You can use a more specific exception (need a `from galaxy.exceptions import RequestParameterInvalidException` )

```suggestion
            raise RequestParameterInvalidException("The current datatype does not allow datatype changes.")
```

simonbray committed:
add 1 more error message, use more specific exception

nsoranzo committed:
Fix use of imported `exceptions` module

It was working by chance because of the presence of another import (`import galaxy.datatypes.metadata`).

nsoranzo commented on lib/galaxy/managers/datasets.py:
```diff
@@ -684,6 +685,20 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
+        if not self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            raise exceptions.RequestParameterInvalidException("The target datatype does not allow datatype changes.")
+        if not DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
+            raise exceptions.RequestParameterInvalidException("Dataset metadata could not be updated, probably because it is not in an 'ok' state.")
```
That's not an invalid parameter in the request, so I would raise a normal exception (I don't think we have any specific Galaxy exception), which should also return a 500 HTTP error code (this is in fact more a server than user error).

```suggestion
            raise Exception("Dataset metadata could not be updated, probably because it is not in an 'ok' state.")
```

nsoranzo commented on lib/galaxy_test/api/test_datasets.py:
```diff
@@ -134,3 +135,30 @@ def test_tag_change(self):
+        invalidly_updated_hda = self._put(  # try updating with invalid datatype
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id),
+            {'datatype': 'invalid'})
+        self._assert_status_code_is(invalidly_updated_hda, 400)
```
```suggestion
        invalidly_updated_hda_response = self._put(  # try updating with invalid datatype
            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id),
            {'datatype': 'invalid'})
        self._assert_status_code_is(invalidly_updated_hda_response, 400)
```

nsoranzo commented on lib/galaxy_test/api/test_datasets.py:
```diff
@@ -134,3 +135,30 @@ def test_tag_change(self):
+            {'datatype': 'tabular'}).status_code
+        assert update_while_incomplete == 400
+
+        time.sleep(10)  # now wait for upload to complete
```
```suggestion
        self.dataset_populator.wait_for_history(self.history_id)  # now wait for upload to complete
```

nsoranzo commented on lib/galaxy_test/api/test_datasets.py:
```diff
@@ -1,4 +1,5 @@
 import textwrap
+import time
```
```suggestion
```

nsoranzo commented on lib/galaxy_test/api/test_datasets.py:
```diff
@@ -134,3 +135,30 @@ def test_tag_change(self):
+        update_while_incomplete = self._put(  # try updating datatype before upload is complete
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id),
+            {'datatype': 'tabular'}).status_code
+        assert update_while_incomplete == 400
```
```suggestion
        update_while_incomplete_response = self._put(  # try updating datatype before upload is complete
            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id),
            {'datatype': 'tabular'})
        self._assert_status_code_is(update_while_incomplete_response, 500)
```

nsoranzo commented on lib/galaxy_test/api/test_datasets.py:
```diff
@@ -134,3 +135,30 @@ def test_tag_change(self):
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id)).json()
+        assert original_hda['extension'] == 'txt'
+        assert original_hda['data_type'] == 'galaxy.datatypes.data.Text'
+        assert len(original_hda['visualizations']) < 10
```
Maybe checking that a tabular-specific visualization is not present in `original_hda['visualizations']` is more future-proof? Same (reversed) below.

simonbray committed:
changes to datatype change test

nsoranzo commented on lib/galaxy_test/api/test_datasets.py:
```diff
@@ -142,23 +141,23 @@ def test_update_datatype(self):
-        assert len(successful_updated_hda['visualizations']) > 20
+        assert successful_updated_hda_response['extension'] == 'tabular'
+        assert successful_updated_hda_response['data_type'] == 'galaxy.datatypes.tabular.Tabular'
+        'scatterplot' not in successful_updated_hda_response['visualizations']
```
Shouldn't this be:
```suggestion
        'scatterplot' in successful_updated_hda_response['visualizations']
```

simonbray replied:
Yes, also it's not actually an assertion... just fixed it, thanks.

simonbray committed:
fix assertion

mvdbeek commented on lib/galaxy/managers/datasets.py:
```diff
@@ -684,6 +683,20 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
+        if not self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            raise exceptions.RequestParameterInvalidException("The target datatype does not allow datatype changes.")
+        if not DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
+            raise Exception("Dataset metadata could not be updated, probably because it is not in an 'ok' state.")
```
```suggestion
            raise RequestParameterInvalidException("Dataset metadata could not be updated because it is used as input or output of a running job.")
```
I would make this a `RequestParameterInvalidException`, this shouldn't be a 500.

simonbray replied:
`used as input or output` is not clear for me - I only encountered this error so far when a dataset is uploading (like in the test) or a previous datatype change is not complete

mvdbeek replied:
I just edited the message, `used as input or output of a running job.` which is what you're describing, right ?

mvdbeek reviewed (approved):
Awesome, that looks great to me!

simonbray committed:
update exception message

mvdbeek commented on lib/galaxy_test/api/test_datasets.py:
```diff
@@ -134,3 +134,30 @@ def test_tag_change(self):
+        update_while_incomplete_response = self._put(  # try updating datatype before upload is complete
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id),
+            {'datatype': 'tabular'})
+        self._assert_status_code_is(update_while_incomplete_response, 500)
```
I think that could lead to unstable API tests as performance increases. We've got a a tool that can sleep before running, I'd use that for this test:
```
        inputs = {
            'input1': {'src': 'hda', 'id': hda_id},
            'sleep_time': 10,
        }
        run_response = self.dataset_populator.run_tool(
            "cat_data_and_sleep",
            inputs,
            history_id,
            assert_ok=False,
        )
        queued_id = run_response["outputs"][0]["id"]
```

mvdbeek commented on lib/galaxy/managers/datasets.py:
```diff
@@ -684,6 +683,20 @@ def deserialize_metadatum(self, dataset_assoc, key, val, **context):
+        if not self.app.datatypes_registry.get_datatype_by_extension(val).allow_datatype_change:
+            raise exceptions.RequestParameterInvalidException("The target datatype does not allow datatype changes.")
+        if not DatasetAssociationManager(self.app).ok_to_edit_metadata(item.dataset_id):
+            raise Exception("Dataset metadata could not be updated because it is used as input or output of a running job.")
```
```suggestion
            raise RequestParameterInvalidException("Dataset metadata could not be updated because it is used as input or output of a running job.")
```
This isn't a server error, I think this falls into the invalid request parameter domain

nsoranzo replied:
I previously suggested the inverse :laughing: . I suppose it's border-line, Galaxy could put the request in a queue instead of returning an error. Also, the dataset may be in use by another user with whom it is shared.

mvdbeek replied:
I agree about the queue, but I think this is nothing admins should see as errors in the logs, which would happen with bare Exceptions, right ?
